dev/PushNotifications/PushNotificationChannel.cpp

```diff
@@ -89,15 +88,35 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
     {
         if (IsPackagedAppScenario())
         {
-            return m_channel.PushNotificationReceived([weak_self = get_weak(), handler](auto&&, auto&& args)
+            if (m_channel)
             {
-                if (auto strong = weak_self.get())
+                return m_channel.PushNotificationReceived([weak_self = get_weak(), handler](auto&&, auto&& args)
                 {
-                    auto pushArgs = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationReceivedEventArgs>(args);
-                    pushArgs.Handled(true);
-                    handler(*strong, pushArgs);
-                };
-            });
+                    if (auto strong = weak_self.get())
+                    {
+                        auto pushArgs = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationReceivedEventArgs>(args);
+                        pushArgs.Handled(true);
+                        handler(*strong, pushArgs);
+                    };
+                });
+            }
+            else
+            {
+                auto lock = m_lock.lock_exclusive();
+
+                THROW_HR_IF_NULL(E_UNEXPECTED, m_channelInfo.channelUri.c_str());
+
+                wil::unique_cotaskmem_string appUserModelId;
+                THROW_IF_FAILED(GetAppUserModelId(appUserModelId));
+
+                THROW_IF_FAILED(PushNotifications_RegisterNotificationSinkForFullTrustApplication(appUserModelId.get(), this));
```

Comment:
You only want to register the foreground sink if the foregroundHandlerCount is 0. Otherwise, we keep the existing sink.

Reply:
question here: should we not always register the latest sink? @hulumane.

Reply:
You misunderstand the design for foreground sink. The way @pmpurifoy  implemented this is using a single sink registration which is refcounted for every eventHandler that the developer subscribes too. 

Another way to achieve the same result is to AddRef the COM ref instead of having a foregroundHandlerCount (we can delete this). That might actually be more optimal.

Reply:
https://github.com/microsoft/WindowsAppSDK/issues/1489 - created a bug for this as this is an improvement and will not impact the current design. Thanks.